docs: Add example of generating tests with a HOF (#4494)
mochajs/mocha: docs/index.md

Take the following example:

```js
const assert = require('chai').assert;

> function add() {

Reviewer: ```suggestion
function add(...args) {
```

>   return Array.prototype.slice.call(arguments).reduce((prev, curr) => prev + curr, 0);

Reviewer: ```suggestion
  return args.reduce((prev, curr) => prev + curr, 0);
```

}

> describe('add()', () => {

Reviewer: ```suggestion
describe('add()', function() {
```

  const tests = [
    {args: [1, 2], expected: 3},
    {args: [1, 2, 3], expected: 6},
    {args: [1, 2, 3, 4], expected: 10}
  ];

  tests.forEach(({args, expected}) => {

>     it('correctly adds ' + args.length + ' args', () => {

Reviewer: ```suggestion
    it(`correctly adds ${args.length} args`, function() {
```

Mocha recommends _not_ using an arrow function for callbacks because of the presence of the context object (`this`).  you _can_ do this if you're not using `this`, but it can lead to subtle bugs in future test maintenance.  best to keep it out of the official docs

    ✓ correctly adds 4 args
```

Tests added inside a `.forEach` handler often don't play well with editor plugins, especially with "right-click run" features.
Another way to parameterize tests is to generate them with a closure. This following example is equivalent to the one above:

```js

> describe('add()', () => {

Reviewer: ```suggestion
describe('add()', function() {
```

>   const testAdd = ({args, expected}) => () => {

Reviewer: ```suggestion
  const testAdd = ({args, expected}) => function() {
```

    ✓ correctly adds 4 args
```

Tests added inside a `.forEach` handler often don't play well with editor plugins, especially with "right-click run" features.

> Another way to paramaterize tests is by generating them with a higher order function. This example is equivalent to the one above.

Reviewer: ```suggestion
Another way to parameterize tests is to generate them with a closure. This following example is equivalent to the one above:
```

Fixed a typo above, but I think "closure" is a more widely-known term in JS

```js
describe('add()', function() {

>   function testAdd(test) {

Reviewer: Can we use more modern syntax here?  Looks better IMO.  :wink:

```suggestion
  function testAdd({args, expected}) {
    const result = add(...args);
    assert.equal(result, expected);
  }
```
